### DISCUSSION TITLE
Every .tool-versions needs a .envrc?

### DISCUSSION THREAD

**Discussion by gullitmiranda** [asdf-community/asdf-direnv#101] [Q&A] [answered]
When we are using the `asdf` standalone, it always takes into consideration the `.tool-versions` of the current directory, however the `use asdf` behaves different.

This takes me the following questions:

1. Is this behavior really expected? If yes, I think it would be good to add a note about it.

2. Is there no way to always take into consideration '.tool-versions' from the current directory? If anyone can help me by identifying `.tool-version` correct I believe I can work on this change

> This difference in behavior made me lose a few hours trying to understand why `.tool-versions` of the sub project did not work 🤦

## My Setup

- `~/.envrc`

```shell
use asdf
```

- `~/.tool-versions`

```
direnv 2.30.3
nodejs 17.4.0
```

- `~/project/.envrc`

```shell
source_up .envrc
use asdf
```

- `~/project/.tool-versions`

```
nodejs 16.14.0
```

- `~/project/sub-project/.tool-versions`

```
nodejs 14.19.0
```

### Behaviors

- (✅ OK) Home directory:

```shell
$ cd ~/
direnv: loading ~/.envrc
direnv: using asdf
direnv: Creating env file ~/.asdf/installs/direnv/2.30.3/env/3953415231-2784736007-4011808011-2769317095
direnv: loading ~/.asdf/installs/direnv/2.30.3/env/3953415231-2784736007-4011808011-2769317095
direnv: using asdf nodejs 17.4.0
direnv: using asdf direnv 2.30.3

$ node --version
v17.4.0

$ which node
~/.asdf/installs/nodejs/17.4.0/bin/node
```

- (✅ OK) Project directory:

```shell
$ cd ~/project
direnv: loading ~/project/.envrc
direnv: loading ~/.envrc
direnv: using asdf
direnv: loading ~/.asdf/installs/direnv/2.30.3/env/3953415231-2784736007-4011808011-2769317095
direnv: using asdf nodejs 17.4.0
direnv: using asdf direnv 2.30.3
direnv: using asdf
direnv: loading ~/.asdf/installs/direnv/2.30.3/env/2059550318-3277659698-894641201-398937083
direnv: using asdf pnpm 6.30.1
direnv: using asdf direnv 2.30.3
direnv: using asdf nodejs 16.14.0

$ node --version
v16.14.0

$ which node
~/.asdf/installs/nodejs/16.14.0/bin/node
```

- (❌ not the same behaviour of `asdf`) Sub Project directory:

```shell
$ cd ~/project/sub-project
# same output from ~/project

$ node --version
v16.14.0

$ which node
~/.asdf/installs/nodejs/16.14.0/bin/node
```

**Comment by jfly** [comment, marked as answer]
Ooh that's a very interesting point, and I can see the confusion. Unfortunately, I don't think we can change this behavior in any sane way, as direnv executes `.envrc` files with cwd set to the folder that script is in. In other words, in your example, from within the `~/project/.envrc` file, there's no way to tell the difference between "we're currently in `~/project`" and "we're currently in `~/project/sub-project/`".

This is actually good and useful behavior for `direnv` to have. If it didn't act that way, you might experience different behavior in this two situations:

1. cd-ing directly from `~` to `~/project/sub-project`
2. cd-ing from `~` to `~/project` and then to `~/project/sub-project`

In scenario 1), you'll end up with an environment formed by evaluating `~/project/.envrc` with CWD = 
 `~/project/sub-project`, and in scenario 2), you'll end up with an environment formed by evaluating  `~/project/.envrc` with CWD = 
 `~/project/` (the second `cd` won't do anything, because there's no new .envrc file to evaluate).

https://github.com/direnv/direnv/issues/218 talks about this exact issue in great detail.

So, I think the answer to your question is: "yes, this behavior is expected". Would you be open to sending in a PR adding a note about it, @gullitmiranda?
